### PR TITLE
Replace asset library with photos library

### LINKED ISSUE #4141: Media problems with iCloud Photo Library
I was trying to add an image to a post that was a bit old, so I started scrolling through my photos on the image picker but couldn't find it. The album selector said that my 'all photos' collection had about 400 items, but the photos app said 2000+. I noticed it had my latest photos then some really old ones that were on my iPad's camera roll. 
I noticed some other albums that showed a thumbnail but 0 images. 

I switched to the photos app, opened that album, and viewed one of the images. I noticed it had to download it from iCloud. When I switched back to the app the photo popped up, but then the app crashed.

I went back to the photos app to find the photo I originally wanted, made sure it was preloaded, and then it was available in the app.

### PR DESCRIPTION
**Pull request by SergioEstevao**
Closes #4141 #3759 

This PR updates/replaces all the AssetLibrary framework code, that is deprecated on iOS9, with the Photos framework.
In the process it also fixes a lot of issues on the Media Picker missing asssets because the Photos framework handles correctly iCloud Photos.

How to test:

Using the editor, try to add images, videos to the post and also set featured images.
If possible try all different combinations of:

- Photos on landscape, portrait, rotated, edited with filters, etc..
- Videos on landscape and portrait, rotated, cropped, etc..
- Photos and Videos saved on iCloud.
- Capture photos and video using the picker.
 
Needs Review: @astralbodies

### CODE REVIEW

**Comment by astralbodies** [comment]
@SergioEstevao it looks like its not merge-able - do you want to resolve this first before I review in case the merge breaks something?

**Comment by SergioEstevao** [comment]
@astralbodies it's now ready to go.
For future reference it looks that now that we applied branch protection on develop, if your tests fail it says non-mergeable. This can be a issue if have tests failing because of external issues. @kwonye was this the goal of activating branch protection?

**Comment by kwonye** [comment]
> This can be a issue if have tests failing because of external issues.

@SergioEstevao unfortunately due to how `develop` was previously broken, there may be instances of this still lingering around. From now on however, `develop` should always be green, so any red should technically be from relevant work in the respective branch.

**Comment by astralbodies** [comment]
Folder name at the top shows (null) during access request:
img_0005

**Comment by astralbodies** [comment]
If possible, provide more information about "trying again later". In most cases this is probably due to a permissions issue - should give instructions on where to enable photo access in Settings.app.
img_0006

**Comment by astralbodies** [comment]
After declining access to my photos, I went out to Settings.app into Privacy and allowed access. I came back into the app, tapped the media button and still got the "There was a problem when trying to access your media" error.

**Comment by astralbodies** [comment]
Tapped the "Camera Roll" popover while in portrait, then rotated to landscape. This is the result and I cannot tap outside the popover to dismiss it:
img_0007

**Comment by astralbodies** [comment]
I got the "There was a problem when trying to access your media" after allowing access to my photo library for the first time on my iPad.
img_0006

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        maximumResolution: CGSize,
+        stripGeoLocation: Bool,
+        successHandler: SuccessHandler,
+        errorHandler: ErrorHandler) {
+        
+        switch self.mediaType {
+        case .Image:
+            exportImageToURL(url,
+                targetUTI: targetUTI,
+                maximumResolution:maximumResolution,
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: Code style - spaces after colon in all method parameters of the function calls in this switch

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        case .Video:
+            exportVideoToURL(url,
+                targetUTI: targetUTI,
+                maximumResolution:maximumResolution,
+                stripGeoLocation:stripGeoLocation,
+                successHandler: successHandler,
+                errorHandler: errorHandler)
+        default:
+            errorHandler(error: errorForCode(.UnsupportedAssetType,
+                failureReason: NSLocalizedString("UnsupportedAssetType", comment: "Error reason to display when exporting an unknow asset type from the device library")))
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: Does this string need to look prettier as English is the base language type? (There was a problem exporting one of your images or videos etc etc).

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        }
+    }
+    
+    func exportImageToURL(url: NSURL,
+        targetUTI: String,
+        maximumResolution:CGSize,
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: Code style - spacing :smile:

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        
+        let options = PHImageRequestOptions()
+        options.version = .Current
+        options.deliveryMode = .HighQualityFormat
+        options.resizeMode = .Exact
+        options.synchronous = false
+        options.networkAccessAllowed = true
+        var requestedSize = maximumResolution
+        if (requestedSize == CGSize.zero) {
+            requestedSize = CGSize(width: self.pixelWidth, height: self.pixelHeight)
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: Is `self` needed here? I know there's a tendency to not use self in Swift as much as possible.

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+                    resultingMetadata[kCGImagePropertyTIFFDictionary as String] = tiffMetadata
+                }
+            }
+        }
+        return resultingMetadata
+    }
+    
+    func exportVideoToURL(url: NSURL,
+        targetUTI: String,
+        maximumResolution:CGSize,
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: Code style - spacing

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+     The targetSize is the maximum resulting resolution  the resultSize will normally be a lower value that mantains the aspect ratio of the asset
+     
+     - Parameters:
+        - url: file url to where the asset should be exported, this must be writable location
+        - targetSize:  the maximum pixel resolution that the file can have after exporting. If CGSizeZero is provided the original size of image is returned.
+        - successHandler: a handler that will be invoked on success with the resulting resolution of the image
+        - errorHandler: a handler that will be invoked when some error occurs when generating the thumbnail
+     */
+    func exportThumbnailToURL(url: NSURL,
+        targetSize:CGSize,
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: Code style - spacing

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+                            failureReason: NSLocalizedString("Unable to export metadata", comment: "Error reason to display when the export of a image from device library fails")
+                            ))
+                    }
+                    return
+            }
+            completionBlock(metadata:image.properties)
+        }
+    }
+    
+    func originalUTI() -> (String?) {
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: Code style - I don't think its common to wrap the return type in parenthesis unless its a tuple

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+        }
+        for resource in resources {
+            if (types.containsObject(resource.type.rawValue) ) {
+                return resource.uniformTypeIdentifier
+            }
+        }
+        return nil
+    }
+    
+    func originalFilename() -> (String?) {
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: Code style - I don't think its common to wrap the return type in parenthesis unless its a tuple

File: WordPress/Classes/Extensions/UIImage+Exporters.swift [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+     Writes an image to a url location with the designated type format and EXIF metadata
+     
+     - Parameters:
+     - url: file url to where the asset should be exported, this must be writable location
+     - type: the UTI format to use when exporting the asset
+     - compressionQuality: defines the compression quality of the export. This is only relevant for type formats that support a quality parameter. Ex: jpeg
+     - metadata: the image metadata to save to file.
+     */
+    func writeToURL(url: NSURL, type:String, compressionQuality :Float = 0.9,  metadata:[String:AnyObject]? = nil) throws -> ()
+    {
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/UIImage+Exporters.swift]: Code style - bracket on the same line as func definition

File: WordPress/Classes/Extensions/UIImage+Exporters.swift [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        let properties: [String:AnyObject] = [kCGImageDestinationLossyCompressionQuality as String: compressionQuality]
+        
+        guard let destination = CGImageDestinationCreateWithURL(url, type, 1, nil),
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/UIImage+Exporters.swift]: Spacing seems a little off here - maybe line up the lets and put else on the same line as the second let

File: WordPress/Classes/Services/MediaService.m [hunk 11]
```diff
@@ -380,44 +403,42 @@ - (Media *)newMediaForPost:(AbstractPost *)post
...
+    return mediaURL;
+}
+
+- (NSURL *)urlForMediaWithFilename:(NSString *)filename andExtension:(NSString *)extension
+{
+    NSURL *mediaDirectoryURL = [self urlForMediaDirectory];
+    NSString *basename= [[filename stringByDeletingPathExtension] lowercaseString];
+    NSURL *resultURL = [mediaDirectoryURL URLByAppendingPathComponent:basename];
+    resultURL = [resultURL URLByAppendingPathExtension:extension];
+    NSUInteger *index = 1;
```
Comment (astralbodies) [inline WordPress/Classes/Services/MediaService.m]: I don't think this is meant to make this a pointer

File: WordPress/Classes/Services/MediaService.m [hunk 12]
```diff
@@ -380,44 +403,42 @@ - (Media *)newMediaForPost:(AbstractPost *)post
...
-    NSUInteger index = 0;
-    while ([fileManager fileExistsAtPath:path]) {
+    
+    return mediaURL;
+}
+
+- (NSURL *)urlForMediaWithFilename:(NSString *)filename andExtension:(NSString *)extension
+{
+    NSURL *mediaDirectoryURL = [self urlForMediaDirectory];
+    NSString *basename= [[filename stringByDeletingPathExtension] lowercaseString];
```
Comment (astralbodies) [inline WordPress/Classes/Services/MediaService.m]: Code style - space before =

**Comment by astralbodies** [comment]
@SergioEstevao when you have the updates made to the comments, ping me and I'll review again. :smile:

File: WordPress/Classes/Extensions/PHAsset+Exporters.swift (hunk 1 above)
Comment (SergioEstevao) [inline WordPress/Classes/Extensions/PHAsset+Exporters.swift]: @astralbodies I don't see anything in our [style guide](https://github.com/wordpress-mobile/swift-style-guide) about this rule and all examples there have the space. I actually think it make the code clear to have that space.

**Comment by SergioEstevao** [comment]
@astralbodies ready for another round of review. All the permission issues are solved, code style issues also the only thing I was unable to reproduce was the media group popup issue: where you couldn't dismiss it after rotation? Do you mind to test it again and if it still happens tell me the steps to reproduce it. Thanks!

**Comment by astralbodies** [comment]
Awesome! :shipit:

**Pull request merged**